docs/DafnyRef/Specifications.md

```diff
@@ -844,3 +844,61 @@ add:
    requires Valid()
    reads Repr
 ```
+
+## 7.8. Well-formedness of specifications {#sec-well-formedness}
+
+Dafny ensures that the [`requires` clauses](#sec-requires-clause) and [`ensures` clauses](#sec-ensures-clauses) are [well-formed](#sec-assertion-batches) independent of the body they belong to. Examples of conditions this rules out are null pointers, out-of-bounds array access, and division by zero.
+Hence, when declaring the following method:
+
+```dafny
+method Test(a: array<int>) returns (j: int)
+  requires a.Length >= 1
+  ensures a.Length % 2 == 0 ==> j >= 10 / a.Length
+{
+  j := 20;
+  var divisor := a.Length;
+  if divisor % 2 == 0 {
+    j := j / divisor;
+  }
+}
+```
+
+Dafny will split the verification in two [assertion batches](#sec-assertion-batches)
+that will roughly look like the following lemmas:
+
+```dafny
+lemma Test_WellFormed(a: array<int>)
+{
+  assert a != null;       // for the `requires a.Length >= 1`
+  assume a.Length >= 1;   // After well-formedness, we assume the requires
+  assert a != null;       // Again for the `a.Length % 2`
+  if a.Length % 2 == 0 {
+    assert a != null;     // Again for the final `a.Length`
+    assert a.Length != 0; // Because of the 10 / a.Length
+  }
+}
+
+lemma Test_Correctness(a: array<int>)
```

Comment:
Is this actually legal within a lemma? Maybe this needs to be a (ghost) method.

Reply:
With a var in front of the j it is legal. but I did convert it to a method.
I also added check-verify.